### PR TITLE
Fix ReverbEffect

### LINKED ISSUE #88: Effect properties don't raise ValueError when set to a value out of range
Several tests are failing currently. They don't raise `ValueError` where they meant to be.

According to [OpenAL doc](https://kcat.strangesoft.net/misc-downloads/Effects%20Extension%20Guide.pdf) of `AL_EFFECT_EAXREVERB`, (which I suppose `EFXEAXREVERBPROPERTIES` implements), then `reverb_density` should be in range `[0, 1]`, but here it doesn't raise any `ValueError` (or any error, for that matter) if I set it to -1 or 2. This is an example, 14 other tests in `test-eff` are failing for the same reason.

Is this what it means to be? Am I looking at the wrong doc?

### PR DESCRIPTION
**Pull request by 9a24f0**

### CODE REVIEW

File: src/palace.pyx [hunk 1]
```diff
@@ -89,6 +89,7 @@ from types import TracebackType
 from typing import (Any, Callable, Dict, Iterable, Iterator,
                     List, Optional, Sequence, Tuple, Type)
 from warnings import catch_warnings, simplefilter, warn
+from numpy import linalg as LA
```
Comment (McSinyx) [inline src/palace.pyx]: no

File: src/palace.pyx [hunk 2]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
     def gain(self, value: float) -> None:
+        if value < 0.0 or value > 1.0:
+            raise ValueError(f'invalid gain: {value}')
         self.properties.gain = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def gain_hf(self) -> float:
+        """Gain HF, from 0.0 to 1.0."""
```
Comment (McSinyx) [inline src/palace.pyx]: It's high frequency gain

File: src/palace.pyx [hunk 3]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.gain_hf
 
     @gain_hf.setter
     def gain_hf(self, value: float) -> None:
+        if value < 0.0 or value > 1.0:
+            raise ValueError(f'invalid gain hf: {value}')
```
Comment (McSinyx) [inline src/palace.pyx]: Same here.

File: src/palace.pyx [hunk 4]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
         self.properties.gain_hf = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def gain_lf(self) -> float:
+        """Gain LF, from 0.0 to 1.0."""
```
Comment (McSinyx) [inline src/palace.pyx]: Here too.

File: src/palace.pyx [hunk 5]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.gain_lf
 
     @gain_lf.setter
     def gain_lf(self, value: float) -> None:
+        if value < 0.0 or value > 1.0:
+            raise ValueError(f'invalid gain lf: {value}')
```
Comment (McSinyx) [inline src/palace.pyx]: aussi

File: src/palace.pyx [hunk 6]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
     def decay_time(self, value: float) -> None:
+        if value < 0.1 or value > 20.0:
+            raise ValueError(f'invalid decay time: {value}')
         self.properties.decay_time = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def decay_hf_ratio(self) -> float:
+        """Decay HF ratio, from 0.1 to 20.0."""
```
Comment (McSinyx) [inline src/palace.pyx]: another one

File: src/palace.pyx [hunk 7]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.decay_hf_ratio
 
     @decay_hf_ratio.setter
     def decay_hf_ratio(self, value: float) -> None:
+        if value < 0.1 or value > 20.0:
+            raise ValueError(f'invalid decay hf ratio: {value}')
```
Comment (McSinyx) [inline src/palace.pyx]: DJ Khalid

File: src/palace.pyx [hunk 8]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
         self.properties.decay_hf_ratio = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def decay_lf_ratio(self) -> float:
+        """Decay LF ratio, from 0.1 to 20.0."""
```
Comment (McSinyx) [inline src/palace.pyx]: Rick Ross

File: src/palace.pyx [hunk 9]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.decay_lf_ratio
 
     @decay_lf_ratio.setter
     def decay_lf_ratio(self, value: float) -> None:
+        if value < 0.1 or value > 20.0:
+            raise ValueError(f'invalid decay lf ratio: {value}')
```
Comment (McSinyx) [inline src/palace.pyx]: Mayback Music

File: src/palace.pyx [hunk 10]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
     def reflections_delay(self, value: float) -> None:
+        if value < 0.0 or value > 0.3:
+            raise ValueError(f'invalid reflections delay: {value}')
         self.properties.reflections_delay = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def reflections_pan(self) -> Vector3:
+        """Reflections pan, 3D vector of magnitude between 0 and 1."""
```
Comment (McSinyx) [inline src/palace.pyx]:
```suggestion
        """Reflections pan as 3D vector of magnitude between 0 and 1."""
```

File: src/palace.pyx [hunk 11]
```diff
@@ -2088,110 +2089,143 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.reflections_pan
 
     @reflections_pan.setter
     def reflections_pan(self, value: Vector3) -> None:
+        if LA.norm(value) < 0 or LA.norm(value) > 1:
+            raise ValueError(f'invalid reflections pan: {value}')
         self.properties.reflections_pan[0] = value[0]
         self.properties.reflections_pan[1] = value[1]
         self.properties.reflections_pan[2] = value[2]
```
Comment (McSinyx) [inline src/palace.pyx]:
```python
x, y, z = value
magnitude = x*x + y*y + z*z
if ...
    raise...
... = x
...
```

File: src/palace.pyx [hunk 12]
```diff
@@ -2200,30 +2234,39 @@ cdef class ReverbEffect(BaseEffect):
...
     def late_reverb_delay(self, value: float) -> None:
+        if value < 0.0 or value > 0.1:
+            raise ValueError(f'invalid late reverb delay: {value}')
         self.properties.late_reverb_delay = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def late_reverb_pan(self) -> Vector3:
+        """Late reverb pan, 3D vector of magnitude between 0 and 1."""
```
Comment (McSinyx) [inline src/palace.pyx]:
```suggestion
        """Late reverb pan as a 3D vector of magnitude between 0 and 1."""
```

File: src/palace.pyx [hunk 13]
```diff
@@ -2200,30 +2234,39 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.late_reverb_pan
 
     @late_reverb_pan.setter
     def late_reverb_pan(self, value: Vector3) -> None:
+        if LA.norm(value) < 0 or LA.norm(value) > 1:
+            raise ValueError(f'invalid late reverb pan: {value}')
         self.properties.late_reverb_pan[0] = value[0]
         self.properties.late_reverb_pan[1] = value[1]
         self.properties.late_reverb_pan[2] = value[2]
```
Comment (McSinyx) [inline src/palace.pyx]: same as above

File: src/palace.pyx [hunk 14]
```diff
@@ -2232,90 +2275,117 @@ cdef class ReverbEffect(BaseEffect):
...
     def modulation_depth(self, value: float) -> None:
+        if value < 0.0 or value > 1.0:
+            raise ValueError(f'invalid modulation depth: {value}')
         self.properties.modulation_depth = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def air_absorption_gain_hf(self) -> float:
+        """Air absorption gain, from 0.892 to 1.0."""
```
Comment (McSinyx) [inline src/palace.pyx]: high frequency*

File: src/palace.pyx [hunk 15]
```diff
@@ -2232,90 +2275,117 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.air_absorption_gain_hf
 
     @air_absorption_gain_hf.setter
     def air_absorption_gain_hf(self, value: float) -> None:
+        if value < 0.892 or value > 1.0:
+            raise ValueError(f'invalid air absorption gain hf: {value}')
```
Comment (McSinyx) [inline src/palace.pyx]: same as above

File: src/palace.pyx [hunk 16]
```diff
@@ -2232,90 +2275,117 @@ cdef class ReverbEffect(BaseEffect):
...
         self.properties.air_absorption_gain_hf = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def hf_reference(self) -> float:
+        """HF reference, from 1000.0 to 20000.0."""
```
Comment (McSinyx) [inline src/palace.pyx]: we the best music

File: src/palace.pyx [hunk 17]
```diff
@@ -2232,90 +2275,117 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.hf_reference
 
     @hf_reference.setter
     def hf_reference(self, value: float) -> None:
+        if value < 1000.0 or value > 20000.0:
+            raise ValueError(f'invalid hf reference: {value}')
```
Comment (McSinyx) [inline src/palace.pyx]: kiss the ring

File: src/palace.pyx [hunk 18]
```diff
@@ -2232,90 +2275,117 @@ cdef class ReverbEffect(BaseEffect):
...
         self.properties.hf_reference = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def lf_reference(self) -> float:
+        """LF reference, from 20.0 to 1000.0."""
```
Comment (McSinyx) [inline src/palace.pyx]: we the best forever

File: src/palace.pyx [hunk 19]
```diff
@@ -2232,90 +2275,117 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.lf_reference
 
     @lf_reference.setter
     def lf_reference(self, value: float) -> None:
+        if value < 20.0 or value > 1000.0:
+            raise ValueError(f'invalid lf reference: {value}')
```
Comment (McSinyx) [inline src/palace.pyx]: I'm on one

File: src/palace.pyx [hunk 20]
```diff
@@ -2232,90 +2275,117 @@ cdef class ReverbEffect(BaseEffect):
...
     def room_rolloff_factor(self, value: float) -> None:
+        if value < 0.0 or value > 10.0:
+            raise ValueError(f'invalid room rolloff factor: {value}')
         self.properties.room_rolloff_factor = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def decay_hf_limit(self) -> bool:
+        """Decay HF limit, either TRUE or FALSE."""
```
Comment (McSinyx) [inline src/palace.pyx]: high frequency*

File: src/palace.pyx [hunk 21]
```diff
@@ -2232,90 +2275,117 @@ cdef class ReverbEffect(BaseEffect):
...
         return self.properties.decay_hf_limit
 
     @decay_hf_limit.setter
     def decay_hf_limit(self, value: bool) -> None:
+        if value != True or value != False:
+            raise ValueError(f'invalid decay hf limit: {value}')
         self.properties.decay_hf_limit = value
```
Comment (McSinyx) [inline src/palace.pyx]: I'm not sure about this, maybe just `self.properties.decay_hf_limit = bool(value)` is enough.

**Review by McSinyx — CHANGES_REQUESTED** [review]

File: src/palace.pyx (hunk 11 above)
Comment (9a24f0) [reply src/palace.pyx]: u forgot the sqrt LOL

Comment (McSinyx) [reply src/palace.pyx]: well yes but actually no

File: src/palace.pyx [hunk 22]
```diff
@@ -2088,235 +2088,306 @@ cdef class ReverbEffect(BaseEffect):
...
+        if value < 0.0 or value > 10.0:
+            raise ValueError(f'invalid room rolloff factor: {value}')
         self.properties.room_rolloff_factor = value
         self.impl.set_reverb_properties(self.properties)
         self.slot.apply_effect(self.impl)
 
     @property
     def decay_hf_limit(self) -> bool:
-        return self.properties.decay_hf_limit
+        """High frequency decay limit, either TRUE or FALSE."""
```
Comment (McSinyx) [inline src/palace.pyx]: Drop the `, either TRUE or FALSE`, the annotation is enough.

**Review by McSinyx — CHANGES_REQUESTED** [review]
Could you please check if the result of `fx.decay_hf_limit` is a `bool`, i.e. replace the existing `assert` with `assert fx.decay_hf_limit is True`, etc.

File: src/palace.pyx [hunk 23]
```diff
@@ -2382,7 +2382,7 @@ cdef class ReverbEffect(BaseEffect):
 
     @property
     def decay_hf_limit(self) -> bool:
-        """High frequency decay limit, either TRUE or FALSE."""
+        """High frequency decay limit."""
         return True if self.properties.decay_hf_limit else False
```
Comment (McSinyx) [inline src/palace.pyx]: Final nit:
```suggestion
        return bool(self.properties.decay_hf_limit)
```

**Comment by McSinyx** [comment]
Niting on the commit message, please use proper casing (GH-*), and do me a favor by skipping the PR ref, since you can find it when you're on the commit page anyway, e.g.
```diff
-Fix ChorusEffect (#91)
+Fix ChorusEffect
```

**Review by McSinyx — APPROVED** [review]

**Pull request merged**